### PR TITLE
Migrate updates to Prometheus receiver from opencensus-service

### PR DESCRIPTION
**Pull request by dinooliva**

### CODE REVIEW

File: receiver/prometheusreceiver/README.md [hunk 1]
```diff
@@ -186,17 +186,24 @@ and `summary`, not all the data points have the same name, there are some specia
...
 
-5. Tags need to handle carefully
+4. Tags need to handle carefully
 
 ScrapeLoop strips out any tag with empty value, however, in OpenCensus, the tag keys is stored separately, we need to able to get all the possible tag keys
 of the same metric family before committing the metric family to the sink.
 
+5. StartTimestamp and values of metrics of cumulative types
+
+In OpenCensus, every metrics of cumulative type is required to have a StartTimestamp, which records when a metric is first recorded, however, Prometheus
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done - also change references to ocagent to otelsvc.

File: receiver/prometheusreceiver/README.md [hunk 2]
```diff
@@ -232,16 +239,27 @@ Counter as described in the [Prometheus Metric Types Document](https://prometheu
...
 
-The corresponding Ocagent Metric will be:
+and the 2nd one:
+```
+# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{method="post",code="200"} 1028
+http_requests_total{method="post",code="400"}    5
+```
+
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done.

File: receiver/prometheusreceiver/README.md [hunk 3]
```diff
@@ -232,16 +239,27 @@ Counter as described in the [Prometheus Metric Types Document](https://prometheu
 > is a cumulative metric that represents a single monotonically increasing counter whose value can only increase or be 
 > reset to zero on restart
 
-It is one of the most simple metric types we can be found in both systems. Examples of Prometheus Counters is as shown 
-below:
+It is one of the most simple metric types we can find in both systems. However, it is a cumulative type of metric,
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done.

File: receiver/prometheusreceiver/README.md [hunk 4]
```diff
@@ -232,16 +239,27 @@ Counter as described in the [Prometheus Metric Types Document](https://prometheu
 > is a cumulative metric that represents a single monotonically increasing counter whose value can only increase or be 
 > reset to zero on restart
 
-It is one of the most simple metric types we can be found in both systems. Examples of Prometheus Counters is as shown 
-below:
+It is one of the most simple metric types we can find in both systems. However, it is a cumulative type of metric, 
+considering with have two continuous scrapes from a target, with the first one as shown below:
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done.

File: receiver/prometheusreceiver/README.md [hunk 5]
```diff
@@ -403,12 +441,19 @@ Prometheus. We have to set this value to `0` instead.
 
 ### Gaugehistogram
 
-This is an undocumented data type, it shall be same as regular [Histogram](#histogram)
+This is an undocumented data type, and it's not supported currently
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done.

File: receiver/prometheusreceiver/README.md [hunk 6]
```diff
@@ -456,10 +514,13 @@ metrics := []*metricspb.Metric{
 
 ```
 
-The major difference between the two systems is that in Prometheus it uses `quantile`, while in OpenCensus `percentile` 
-is used. Other than that, OpenCensus has optional values for `Sum` and `Count` of a snapshot, however, they are not 
+There's also some differences between the two systems. One of them is that in Prometheus it uses `quantile`, while in OpenCensus it uses `percentile`.
+Other than that, OpenCensus has optional values for `Sum` and `Count` of a snapshot, however, they are not 
 provided in Prometheus, and `nil` will be used for these values.
 
+Other than that, in some prometheus implementations, such as the Python version, Summary is allowed to have no quantiles, in a case
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done.

File: receiver/prometheusreceiver/README.md [hunk 7]
```diff
@@ -186,17 +186,24 @@ and `summary`, not all the data points have the same name, there are some specia
...
 
-4. Group complex metrics such as histogram together in proper order
+3. Group complex metrics such as histogram together in proper order
 
 In Prometheus, a single aggregated type of metric data such as `histogram` and `summary` is represent by multiple metric data points, such as
 buckets and quantiles as well as the additional `_sum` and `_count` data. ScrapeLoop will feed them into the appender individually. The ocagent
 appender need to have a way to bundle them together to transform them into a single Metric Datapoint Proto object. 
 
-5. Tags need to handle carefully
+4. Tags need to handle carefully
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done.

File: receiver/prometheusreceiver/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Modified.

File: receiver/prometheusreceiver/README.md [hunk 8]
```diff
@@ -232,16 +239,27 @@ Counter as described in the [Prometheus Metric Types Document](https://prometheu
...
-The corresponding Ocagent Metric will be:
+and the 2nd one:
+```
+# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{method="post",code="200"} 1028
+http_requests_total{method="post",code="400"}    5
+```
+
+The Prometheus Receiver will only produce one Metric from the 2nd scrape and subsequent ones if any, the 1st scrape,
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Done.

File: receiver/prometheusreceiver/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Modified.

File: receiver/prometheusreceiver/internal/metricfamily.go [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+package internal
+
+import (
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/scrape"
+	"sort"
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/internal/metricfamily.go]: Acknowledged.

File: go.mod [hunk 10]
```diff
@@ -7,27 +7,24 @@ require (
 	contrib.go.opencensus.io/exporter/prometheus v0.1.0
 	contrib.go.opencensus.io/exporter/zipkin v0.1.1
 	contrib.go.opencensus.io/resource v0.1.1
-	github.com/VividCortex/gohistogram v1.0.0 // indirect
 	github.com/apache/thrift v0.0.0-20161221203622-b2a4d4ae21c7
 	github.com/aws/aws-sdk-go v1.19.18 // indirect
-	github.com/bmizerany/perks v0.0.0-20141205001514-d9a9656a3a4b // indirect
 	github.com/census-instrumentation/opencensus-proto v0.2.1
+	github.com/census-instrumentation/opencensus-service v0.1.0
```
[the remark this answers is not in the document]
Comment (dinooliva) [reply go.mod]: Fixed

**Comment by dinooliva** [comment]
Thanks @pjanotti - I've rebased and fixed issues from the new static checks, should be good to go.

**Comment by dinooliva** [comment]
Thanks! When do you think this can be merged?

**Comment by pjanotti** [comment]
Merged it is, thanks @dinooliva.

**Pull request merged**